### PR TITLE
perf: refine onnx implementation

### PR DESCRIPTION
**Pull request by larme**
- multiple outputs support
- more detailed input type casting

## What does this PR address?

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/bentoml/_internal/frameworks/onnx.py [hunk 1]
```diff
@@ -289,21 +296,24 @@ def forward(self, x, bias):
...
+        if isinstance(run_sig, dict)
+        else t.cast(ModelSignatureDict, attr.asdict(run_sig))
+    )
 
-    if len(model.graph.output) > 1:
-        logger.warning(
-            "The model you are attempting to save has more than one output. The ONNX runner will only return the first output."
-        )
+    run_sig_dict["input_spec"] = [MessageToDict(inp) for inp in model.graph.input]
+    run_sig_dict["output_spec"] = [MessageToDict(out) for out in model.graph.output]
```
Comment (larme) [inline src/bentoml/_internal/frameworks/onnx.py:307]: Here I'm not sure if we want to use `input_spec`/`output_spec` field now or we want to reserve these fields until we have unified io spec for runner interface and use some temporary fields to save onnx spec dict for now.

Comment (larme) [reply src/bentoml/_internal/frameworks/onnx.py:307]: but these specs really belongs to `input_spec/output_spec`

**Comment by larme** [comment]
will add api v1 backward computability and more tests later

File: src/bentoml/_internal/utils/onnx.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
```
Comment (aarnphm) [inline src/bentoml/_internal/utils/onnx.py:1]: should we move this file to `src/bentoml/_internal/frameworks/utils/onnx.py`?

File: src/bentoml/_internal/frameworks/onnx.py [hunk 3]
```diff
@@ -32,11 +33,14 @@
...
 
     ProvidersType = list[str | tuple[str, dict[str, t.Any]]]
-    ONNXArgType = ext.NpNDArray | ext.PdDataFrame | torch.Tensor | tf_ext.Tensor
+    from ..utils.onnx import ONNXArgType
+    from ..utils.onnx import ONNXArgCastedType
 
 
 try:
     import onnx
+    from google.protobuf.json_format import MessageToDict
```
[the remark this answers is not in the document]
Comment (larme) [reply src/bentoml/_internal/frameworks/onnx.py:42]: I guess so. It's only needed when saving the model

File: src/bentoml/_internal/utils/onnx.py (hunk 2 above)
Comment (aarnphm) [reply src/bentoml/_internal/utils/onnx.py:1]: Can you also make a PR that move the tensorflow utils to this folder after this PR?

Comment (larme) [reply src/bentoml/_internal/utils/onnx.py:1]: That's what I'm thinking. I will move both to the same folder in another pr

File: src/bentoml/_internal/frameworks/onnx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (larme) [reply src/bentoml/_internal/frameworks/onnx.py:42]: seems that importing onnx will automatically import `google.protobuf`, so the benefit of lazy loading protobuf alone is tiny

File: src/bentoml/_internal/utils/onnx.py (hunk 2 above)
Comment (aarnphm) [reply src/bentoml/_internal/utils/onnx.py:1]: Is this ready to review?

File: src/bentoml/_internal/frameworks/onnx.py (hunk 1 above)
Comment (larme) [reply src/bentoml/_internal/frameworks/onnx.py]: move to ONNXOptions for now

File: src/bentoml/_internal/utils/onnx.py (hunk 2 above)
Comment (larme) [reply src/bentoml/_internal/utils/onnx.py:1]: @aarnphm ready now

File: src/bentoml/_internal/frameworks/onnx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/onnx.py:36]: noted.

File: src/bentoml/_internal/frameworks/onnx.py [hunk 4]
```diff
@@ -321,6 +331,25 @@ def get_runnable(bento_model: bentoml.Model) -> t.Type[bentoml.Runnable]:
...
     """
 
+    # backward compatibility for v1, load raw model to infer
+    # input_specs/output_specs for onnx model
+    if bento_model.info.api_version == "v1":
+
+        raw_model: onnx.ModelProto | None = None
+        options = t.cast(ONNXOptions, bento_model.info.options)
+
+        if not options.input_specs:
```
Comment (aarnphm) [inline src/bentoml/_internal/frameworks/onnx.py:341]: in v1 this will result in attribute error right? because input_specs is not there?

**Review by aarnphm — APPROVED** [review]

**Pull request merged**